fix: Add AWS Bedrock Extended Thinking support for Claude models (#4343, #4201)

hannesrudolph opened the pull request:
## Description

Fixes #4343
Fixes #4201

This PR adds support for AWS Bedrock Extended Thinking (reasoning mode) for Claude 3.7/4 models, with improved error handling and retry logic.

## Changes Made

### 1. Added Extended Thinking Support
- Updated model definitions to include `supportsReasoningBudget: true` for:
  - Claude 3.7 Sonnet (`anthropic.claude-3-7-sonnet-20250219-v1:0`)
  - Claude 4 Sonnet (`anthropic.claude-sonnet-4-20250514-v1:0`)
  - Claude 4 Opus (`anthropic.claude-opus-4-20250514-v1:0`)

### 2. Implemented Reasoning Configuration
- Extended thinking is **disabled by default** (as per AWS requirements)
- Only enabled when user explicitly sets `enableReasoningEffort: true`
- Added `additionalModelRequestFields` with thinking configuration when enabled
- Uses correct `anthropic_version: "bedrock-20250514"`
- Removes incompatible parameters (temperature, topP) when thinking is enabled

### 3. Enhanced Stream Processing
- Added support for new event types:
  - `thinking` content blocks
  - `thinking_delta` events
  - `signature_delta` events
- Properly yields reasoning chunks with `type: "reasoning"`

### 4. Improved Error Handling
- Added retry logic with exponential backoff (up to 3 retries)
- Added specific error type for thinking-related errors
- Improved generic error message from "Unknown Error" to more descriptive text
- Errors now properly parsed and displayed to users

### 5. Updated UI
- Added ThinkingBudget component to Bedrock settings
- Shows "Enable Reasoning Mode" toggle for supported models
- Allows configuration of max thinking tokens (1024-80% of max output tokens)

## Testing

- [x] All existing tests pass
- [x] Added comprehensive test suite for reasoning functionality
- [x] Manual testing completed:
  - Verified reasoning is disabled by default
  - Verified reasoning only enabled when explicitly requested
  - Tested error handling and retry logic
  - Confirmed UI shows reasoning toggle only for supported models

## Verification of Acceptance Criteria

- [x] "Enable reasoning" option now available for Bedrock Claude models
- [x] Better error handling with retry logic (addresses @hannesrudolph's concern)
- [x] Improved error parsing and messages
- [x] Support for Claude 3.7/4 extended thinking via AWS Bedrock
- [x] Reasoning is opt-in only (disabled by default)

## Checklist

- [x] Code follows project style guidelines
- [x] Self-review completed
- [x] Comments added for complex logic
- [x] Documentation updated (if needed)
- [x] No breaking changes
- [x] Tests added for new functionality

## Additional Context

This implementation follows the technical requirements where extended thinking/reasoning is disabled by default and only enabled when explicitly requested by the user. This maintains backward compatibility while adding the new functionality requested in the issues.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -280,20 +335,59 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
+				modelId: modelConfig.id,
+				budgetTokens: params.reasoningBudget,
+			})
+		}
```
Consider extracting this complex reasoning enablement logic into a helper method like `shouldEnableExtendedThinking()`. The multiple nested conditions make this section harder to read and test in isolation.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -280,20 +335,59 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
 
-		const payload = {
+		// Build the base payload
+		const payload: any = {
```
Could we replace `payload: any` with a more specific type? This would improve type safety and make the code more maintainable. Consider creating a `BedrockPayload` interface that includes the optional `anthropic_version` and `additionalModelRequestFields` properties.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -280,20 +335,59 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
+			params.reasoningBudget
+		) {
+			// Add the anthropic_version field required for extended thinking
+			payload.anthropic_version = "bedrock-20250514"
```
Consider extracting this constant to a module-level variable like `BEDROCK_ANTHROPIC_VERSION = "bedrock-20250514"`. This makes it easier to update if AWS changes the version requirement and improves maintainability.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -280,20 +335,59 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
+
+			// Remove temperature, topP, and top_k when thinking is enabled as they are incompatible
+			delete inferenceConfig.temperature
+			delete inferenceConfig.topP
```
Is there documentation about why `temperature` and `topP` are incompatible with thinking mode? This seems like an AWS Bedrock limitation that would be helpful to document in a comment for future maintainers.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -397,21 +491,59 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
+						continue
 					}
-					continue
 				}
```
This stream processing logic has grown quite complex. Consider using a strategy pattern or extracting event handlers into separate methods (e.g., `handleThinkingEvent()`, `handleTextEvent()`) to improve readability and testability.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -792,6 +931,35 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
+	/**
+	 * Preserves thinking blocks from previous assistant messages for tool use continuity
+	 */
+	private preserveThinkingBlocks(messages: Message[]): Message[] {
```
This entire `preserveThinkingBlocks` method appears to be unnecessary complexity. Looking at the implementation, it always returns messages unchanged regardless of the conditions. No other providers (including the Anthropic provider that also supports thinking) have similar logic. 

Could you clarify what specific problem this is trying to solve? If it's not actually needed, I'd recommend removing this method and the call to it on line 650 to simplify the code.

daniel-lxs reviewed (commented):
I can't actually test if this actually works but, I left some suggestions based on how the Anthropic provider works.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -119,6 +149,80 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
+	/**
+	 * Handles thinking content block events
+	 */
+	private *handleThinkingContentBlock(contentBlock: any): Generator<any, void, unknown> {
```
Could we improve type safety here by replacing `any` with a more specific type? Consider defining a `StreamChunk` interface with `type: 'reasoning' | 'text'` and `text: string` properties. This would make the generator return type more explicit and catch potential type errors.

hannesrudolph commented:
## Review Issues Addressed ✅

Fixed the remaining type safety issue:
- Added `StreamChunk` interface with proper `type: 'reasoning' | 'text'` typing
- Updated all generator methods to use `Generator<StreamChunk, void, unknown>` instead of `any`

All other review feedback was already implemented. Tests pass, no lint errors.

hannesrudolph commented:
## Type Safety Fix Applied ✅

Fixed the CI test failures by resolving type compatibility issues:

- **Replaced custom `StreamChunk` interface** with proper `ApiStreamChunk` import
- **Updated all generator method signatures** to use `ApiStreamChunk` type
- **Resolved `yield*` delegation type mismatches** between generator methods

All local tests pass and TypeScript compilation is clean. The CI failures should now be resolved.

daniel-lxs commented on src/api/providers/bedrock.ts:
```diff
@@ -119,6 +152,80 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
+	/**
+	 * Handles thinking content block events
+	 */
+	private *handleThinkingContentBlock(contentBlock: any): Generator<ApiStreamChunk, void, unknown> {
```
The current implementation uses multiple separate handler methods (handleThinkingContentBlock, handleTextContentBlock, handleThinkingDelta, handleSignatureDelta, handleTextDelta) which creates some code duplication and complexity. Consider consolidating these into a unified stream processor for better maintainability:

```typescript
private *processStreamEvent(streamEvent: StreamEvent): Generator<StreamChunk, void, unknown> {
  if (streamEvent.contentBlockStart) {
    const { contentBlock, start } = streamEvent.contentBlockStart
    
    if (contentBlock?.type === "thinking" && contentBlock.thinking !== undefined) {
      yield { type: "reasoning", text: contentBlock.thinking || "" }
    } else if (start?.text || contentBlock?.text) {
      yield { type: "text", text: start?.text || contentBlock?.text || "" }
    }
  }
  
  if (streamEvent.contentBlockDelta?.delta) {
    const { delta } = streamEvent.contentBlockDelta
    
    if (delta.type === "thinking_delta" && delta.thinking) {
      yield { type: "reasoning", text: delta.thinking }
    } else if (delta.type === "signature_delta" && delta.signature) {
      yield { type: "reasoning", text: delta.signature }
    } else if (delta.text) {
      yield { type: "text", text: delta.text }
    }
  }
}
```
